maestrowf/interfaces/script/slurmscriptadapter.py

```diff
@@ -140,10 +140,14 @@ def get_header(self, step):
         procs = resources.get("procs")
         nodes = resources.get("nodes")
         if not procs and not nodes:
-            err_msg = "No explicit resources specified in {}. At least one of " 
-            "\"procs\" or \"nodes\" must be set to a non-zero value.".format(step.name)
+            err_msg = \
+                'No explicit resources specified in {}. At least one' \
+                ' of "procs" or "nodes" must be set to a non-zero' \
+                ' value.'.format(step.name)
             LOGGER.error(err_msg)
             raise RuntimeError(err_msg)
+        elif procs or (procs and not nodes):
```

Comment:
This conditional is slightly incorrect -- you need to check if `procs` is in the `batch` section, not if it's non-zero.

Reply:
You left a comment "This conditional is slightly incorrect -- you need to check if procs is in the batch section, not if it's non-zero."
In regards to this comment...

From my notes, the goal of this conditional check is because SBATCH ntasks is not present even when 'procs: 4' is in the step. So if procs is in the step, doesn't that mean it won't be in the batch section. In addition, procs is obtained via 'resources.get('procs') and resources is composed of batch and run dicts.

Reply:
> From my notes, the goal of this conditional check is because SBATCH ntasks is not present even when 'procs: 4' is in the step. So if procs is in the step, doesn't that mean it won't be in the batch section. In addition, procs is obtained via 'resources.get('procs') and resources is composed of batch and run dicts.

The `procs` key can be in both the batch and the step. So the only time that `procs` part of the conditional will be false is if it's not set in either.

So I noticed that in the `__init__` you also check for the `procs` in the `batch` section. With this conditional here, you will add `procs` to the header twice. I don't think SLURM will throw an error, but the header should be as clean as possible.

In `__init__` instead of adding the `--ntasks` key, add it here. Simply save the `--ntasks` header line in the `__init__` so that it's in the same place as everything else. Something like:
```
def __init__(self, **kwargs):
    ...
    ntask_header = "#SBATCH --ntasks={procs}"
    ...
```

A suggestion on the exception: move to be right after the chainmap construction -- that'll make it so the header throws the exception without the extra logic. If that exception case is try, we can't do anything. 

After we do the loop for adding items to the header, `nodes` will be there if it's specified (it's set to `""` by default) otherwise we won't have `procs` in the header at all at this point. This is now where you should check for both the hint (`procs in self._batch`) or if `nodes` is not set (because we need to give SLURM some sort of resource to use). In theory, you can even move the `exclusive` addition to be later as well so that you're handling resources in one block.

Does that make sense?

Reply:
With the way that the code is structured right now, if procs is not specified in the `batch.info`, it will never be in the `self._batch`. This means that it cannot be added to the header because the condition is not met.